### ISSUE TITLE
When starting up, recovery of shards takes up to 50 minutes

### ISSUE THREAD

**Issue by bluelu** [elastic/elasticsearch#6372]
We have about 1000 indexes, 20k shards and over 400 nodes in our cluster.

When we restart the cluster, it takes about 50 minutes to reach yellow state. All nodes (including the master node) seem to be idling.

It doesn't seem to be traffic related (the clsuter state is only sent more often later when yellow state is being reached).

The master node seems to only request shard recovery every 11 minutes (but not for all shards), causing the long wait.

[2014-05-27 13:41:53,300][DEBUG][index.gateway            ] [server5N5] [2013.06.24.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#1]}
[2014-05-27 13:41:53,325][DEBUG][index.gateway            ] [server5N5] [2014.03.13.0000_000][1] starting recovery from local ... {elasticsearch[server5N5][generic][T#4]}
[2014-05-27 13:41:53,359][DEBUG][index.gateway            ] [server5N5] [2012.07.06.0000_000][5] starting recovery from local ... {elasticsearch[server5N5][generic][T#5]}
[2014-05-27 13:41:53,367][DEBUG][index.gateway            ] [server5N5] [2013.03.30.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#6]}
[2014-05-27 13:41:53,374][DEBUG][index.gateway            ] [server5N5] [2012.05.27.0000_000][0] starting recovery from local ... {elasticsearch[server5N5][generic][T#7]}
[2014-05-27 13:41:53,462][DEBUG][index.gateway            ] [server5N5] [2012.09.05.0000_000][0] starting recovery from local ... {elasticsearch[server5N5][generic][T#8]}
[2014-05-27 13:41:53,469][DEBUG][index.gateway            ] [server5N5] [2013.06.17.0000_000][2] starting recovery from local ... {elasticsearch[server5N5][generic][T#9]}
[2014-05-27 13:41:53,552][DEBUG][index.gateway            ] [server5N5] [2012.05.05.0000_000][9] starting recovery from local ... {elasticsearch[server5N5][generic][T#10]}
[2014-05-27 13:41:53,635][DEBUG][index.gateway            ] [server5N5] [2012.05.22.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#10]}
[2014-05-27 13:41:53,642][DEBUG][index.gateway            ] [server5N5] [2014.01.26.0000_000][0] starting recovery from local ... {elasticsearch[server5N5][generic][T#11]}
[2014-05-27 13:41:53,648][DEBUG][index.gateway            ] [server5N5] [2012.11.09.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#12]}
[2014-05-27 13:41:53,678][DEBUG][index.gateway            ] [server5N5] [2013.12.26.0000_000][1] starting recovery from local ... {elasticsearch[server5N5][generic][T#13]}
[2014-05-27 13:41:53,685][DEBUG][index.gateway            ] [server5N5] [2014.01.11.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#14]}
[2014-05-27 13:41:53,693][DEBUG][index.gateway            ] [server5N5] [2013.11.04.0000_000][1] starting recovery from local ... {elasticsearch[server5N5][generic][T#15]}
[2014-05-27 13:41:53,715][DEBUG][index.gateway            ] [server5N5] [2014.02.18.0000_000][3] starting recovery from local ... {elasticsearch[server5N5][generic][T#16]}
[2014-05-27 13:41:53,739][DEBUG][index.gateway            ] [server5N5] [2012.06.23.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#17]}
[2014-05-27 13:41:53,747][DEBUG][index.gateway            ] [server5N5] [2013.07.07.0000_000][8] starting recovery from local ... {elasticsearch[server5N5][generic][T#18]}
[2014-05-27 13:41:53,754][DEBUG][index.gateway            ] [server5N5] [2013.05.12.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#19]}
[2014-05-27 13:41:53,804][DEBUG][index.gateway            ] [server5N5] [2013.02.15.0000_000][7] starting recovery from local ... {elasticsearch[server5N5][generic][T#20]}
[2014-05-27 13:41:53,995][DEBUG][index.gateway            ] [server5N5] [2013.09.14.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#21]}
[2014-05-27 13:41:54,001][DEBUG][index.gateway            ] [server5N5] [2012.11.04.0000_000][5] starting recovery from local ... {elasticsearch[server5N5][generic][T#22]}
[2014-05-27 13:41:54,007][DEBUG][index.gateway            ] [server5N5] [2013.11.17.0000_000][3] starting recovery from local ... {elasticsearch[server5N5][generic][T#23]}
[2014-05-27 13:41:54,014][DEBUG][index.gateway            ] [server5N5] [2014.03.15.0000_000][9] starting recovery from local ... {elasticsearch[server5N5][generic][T#24]}
[2014-05-27 13:41:54,088][DEBUG][index.gateway            ] [server5N5] [2014.02.17.0000_000][7] starting recovery from local ... {elasticsearch[server5N5][generic][T#25]}
[2014-05-27 13:41:54,095][DEBUG][index.gateway            ] [server5N5] [2013.08.05.0000_000][9] starting recovery from local ... {elasticsearch[server5N5][generic][T#26]}
[2014-05-27 13:41:54,500][DEBUG][index.gateway            ] [server5N5] [2013.07.29.0000_000][0] starting recovery from local ... {elasticsearch[server5N5][generic][T#27]}
[2014-05-27 13:41:54,507][DEBUG][index.gateway            ] [server5N5] [2014.04.02.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#28]}
[2014-05-27 13:41:54,514][DEBUG][index.gateway            ] [server5N5] [2013.11.23.0000_000][7] starting recovery from local ... {elasticsearch[server5N5][generic][T#29]}
[2014-05-27 13:41:54,520][DEBUG][index.gateway            ] [server5N5] [2013.01.04.0000_000][5] starting recovery from local ... {elasticsearch[server5N5][generic][T#30]}
[2014-05-27 13:41:54,526][DEBUG][index.gateway            ] [server5N5] [2013.05.15.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#31]}
[2014-05-27 13:53:33,567][DEBUG][index.gateway            ] [server5N5] [2012.12.13.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#307]}
[2014-05-27 13:53:33,645][DEBUG][index.gateway            ] [server5N5] [2014.01.19.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#296]}
[2014-05-27 13:53:33,732][DEBUG][index.gateway            ] [server5N5] [2014.02.06.0000_000][4] starting recovery from local ... {elasticsearch[server5N5][generic][T#295]}
[2014-05-27 13:53:33,760][DEBUG][index.gateway            ] [server5N5] [2012.05.12.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#294]}
[2014-05-27 13:53:33,816][DEBUG][index.gateway            ] [server5N5] [2013.01.09.0000_000][0] starting recovery from local ... {elasticsearch[server5N5][generic][T#313]}
[2014-05-27 13:53:34,005][DEBUG][index.gateway            ] [server5N5] [2013.06.05.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#288]}
[2014-05-27 13:53:34,099][DEBUG][index.gateway            ] [server5N5] [2012.11.25.0000_000][2] starting recovery from local ... {elasticsearch[server5N5][generic][T#257]}
[2014-05-27 13:53:34,257][DEBUG][index.gateway            ] [server5N5] [2012.09.23.0000_000][3] starting recovery from local ... {elasticsearch[server5N5][generic][T#303]}
[2014-05-27 13:53:34,821][DEBUG][index.gateway            ] [server5N5] [2012.06.21.0000_000][6] starting recovery from local ... {elasticsearch[server5N5][generic][T#304]}

**Comment by ywelsch-t** [comment]
Hey @bleskes, I'm working with @bluelu and @miccon

After some more analysis, there are two reasons why initial shard allocation is slow for us. First, the `LocalGatewayAllocator` is slow. It will hopefully be adressed as part of #9502. Second, the `BalancedShardsAllocator` is slow. In the following, I will detail why this is the case for our setup.

Assume nodes of type A and B (could be SSD and HDD nodes for example). The nodes are marked as A or B through the attribute `node_type`. We now want to ensure that some of the indices are only allocated on A nodes and others only on B nodes. This is realized by setting `index.routing.allocation.require.node_type` either to A or B.

The benchmark appended below (adapted from the class `ClusterAllocationRerouteBenchmark`) applies the `BalancedShardsAllocator` to the cluster setup that was described here. It shows the following outputs: Full allocation takes 17 minutes and allocating a new index takes 30 seconds.

In our actual setup (nodes of type A,B,C,D) things are even worse (>50 minutes total and >40 seconds for new shards).

```
package org.elasticsearch.benchmark.cluster;

import com.google.common.collect.ImmutableMap;
import org.elasticsearch.cluster.ClusterName;
import org.elasticsearch.cluster.ClusterState;
import org.elasticsearch.cluster.metadata.IndexMetaData;
import org.elasticsearch.cluster.metadata.MetaData;
import org.elasticsearch.cluster.node.DiscoveryNodes;
import org.elasticsearch.cluster.routing.RoutingTable;
import org.elasticsearch.cluster.routing.allocation.AllocationService;
import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
import org.elasticsearch.common.logging.ESLogger;
import org.elasticsearch.common.logging.Loggers;
import org.elasticsearch.common.settings.ImmutableSettings;
import org.elasticsearch.common.unit.TimeValue;
import org.elasticsearch.test.ElasticsearchAllocationTestCase;

import java.util.Random;

import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;

public class SlowClusterAllocationRerouteIssue {

    private static final ESLogger logger = Loggers.getLogger(SlowClusterAllocationRerouteIssue.class);
    
    private static final String NODE_TYPE = "node_type";
    
    public static void main(String[] args) {
        final int numberOfRuns = 1;        
        
        final int numAIndices = 1000;
        final int numAShards = 10;
        final int numAReplicas = 1;
        
        final int numBIndices = 3000;
        final int numBShards = 1;
        final int numBReplicas = 1;
        
        final int numberOfANodes = 300;
        final int numberOfBNodes = 300;
        
        AllocationService strategy = ElasticsearchAllocationTestCase.createAllocationService(ImmutableSettings.builder()
                .put("cluster.routing.allocation.node_initial_primaries_recoveries", "30")
                .put("cluster.routing.allocation.node_concurrent_recoveries", "1")
                .build(), new Random(1));

        MetaData.Builder mb = MetaData.builder();
        for (int i = 1; i <= numAIndices; i++) {
            mb.put(IndexMetaData.builder("a_index" + i).settings(
                    ImmutableSettings.settingsBuilder().put("index.routing.allocation.require." + NODE_TYPE, "A"))
                    .numberOfShards(numAShards).numberOfReplicas(numAReplicas));
        }
        for (int i = 1; i <= numBIndices; i++) {
            mb.put(IndexMetaData.builder("b_index" + i).settings(
                    ImmutableSettings.settingsBuilder().put("index.routing.allocation.require." + NODE_TYPE, "B"))
                    .numberOfShards(numBShards).numberOfReplicas(numBReplicas));
        }
        MetaData metaData = mb.build();
        
        RoutingTable.Builder rb = RoutingTable.builder();
        for (int i = 1; i <= numAIndices; i++) {
            rb.addAsNew(metaData.index("a_index" + i));
        }
        for (int i = 1; i <= numBIndices; i++) {
            rb.addAsNew(metaData.index("b_index" + i));
        }
        RoutingTable routingTable = rb.build();
        
        DiscoveryNodes.Builder nb = DiscoveryNodes.builder();
        for (int i = 1; i <= numberOfANodes; i++) {
            nb.put(ElasticsearchAllocationTestCase.newNode("a_node" + i, ImmutableMap.of(NODE_TYPE, "A")));
        }
        for (int i = 1; i <= numberOfBNodes; i++) {
            nb.put(ElasticsearchAllocationTestCase.newNode("b_node" + i, ImmutableMap.of(NODE_TYPE, "B")));
        }
        
        ClusterState initialClusterState = ClusterState.builder(ClusterName.DEFAULT).metaData(metaData).routingTable(routingTable).nodes(nb).build();
        
        long start = System.currentTimeMillis();
        for (int i = 0; i < numberOfRuns; i++) {
            logger.info("[{}] starting... ", i);
            long runStart = System.currentTimeMillis();
            ClusterState clusterState = initialClusterState;
            while (clusterState.readOnlyRoutingNodes().hasUnassignedShards()) {
                logger.info("[{}] remaining unassigned {}", i, clusterState.readOnlyRoutingNodes().unassigned().size());
                RoutingAllocation.Result result = strategy.applyStartedShards(clusterState, clusterState.readOnlyRoutingNodes().shardsWithState(INITIALIZING));
                clusterState = ClusterState.builder(clusterState).routingResult(result).build();
                result = strategy.reroute(clusterState);
                clusterState = ClusterState.builder(clusterState).routingResult(result).build();
            }
            logger.info("[{}] took {}", i, TimeValue.timeValueMillis(System.currentTimeMillis() - runStart));
        }
        long took = System.currentTimeMillis() - start;
        logger.info("total took {}, AVG {}", TimeValue.timeValueMillis(took), TimeValue.timeValueMillis(took / numberOfRuns));
    }
}
```


Output:

```
[2015-03-06 16:07:28,080][INFO ][org.elasticsearch.benchmark.cluster] [0] starting... 
[2015-03-06 16:07:28,119][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 26000
[2015-03-06 16:07:42,633][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 14000
[2015-03-06 16:08:12,875][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 12700
[2015-03-06 16:08:42,247][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 12100
[2015-03-06 16:09:11,950][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 11500
[2015-03-06 16:09:42,086][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 10900
[2015-03-06 16:10:12,184][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 10300
[2015-03-06 16:10:42,527][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 9700
[2015-03-06 16:11:13,257][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 9100
[2015-03-06 16:11:44,172][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 8500
[2015-03-06 16:12:15,184][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7900
[2015-03-06 16:12:46,170][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7300
[2015-03-06 16:13:16,607][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7000
[2015-03-06 16:13:45,755][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6700
[2015-03-06 16:14:14,863][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6400
[2015-03-06 16:14:43,883][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6100
[2015-03-06 16:15:12,949][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5800
[2015-03-06 16:15:41,760][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5500
[2015-03-06 16:16:10,331][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5200
[2015-03-06 16:16:39,046][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4900
[2015-03-06 16:17:07,332][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4600
[2015-03-06 16:17:35,698][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4300
[2015-03-06 16:18:03,819][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4000
[2015-03-06 16:18:31,773][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3700
[2015-03-06 16:18:59,798][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3400
[2015-03-06 16:19:27,735][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3100
[2015-03-06 16:19:55,748][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2800
[2015-03-06 16:20:23,786][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2500
[2015-03-06 16:20:51,744][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2200
[2015-03-06 16:21:19,735][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1900
[2015-03-06 16:21:47,910][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1600
[2015-03-06 16:22:16,249][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1300
[2015-03-06 16:22:44,494][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1000
[2015-03-06 16:23:12,843][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 700
[2015-03-06 16:23:40,748][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 400
[2015-03-06 16:24:08,520][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 200
[2015-03-06 16:24:26,921][INFO ][org.elasticsearch.benchmark.cluster] [0] took 16.9m
[2015-03-06 16:24:26,922][INFO ][org.elasticsearch.benchmark.cluster] total took 16.9m, AVG 16.9m
```

Note that with some crude hacks (having BalancedShardsAllocator account for the different node types in earlier pathes of the code) we reduced full allocation to a minute and allocation of new shards to a second.

**Comment by bleskes** [comment]
Thx @ywelsch-t . This is a great reproduction. We'll look at it as soon as possible.

**Comment by bleskes** [comment]
@ywelsch-t quick question in the meantime, can you elaborate on what you did:

> Note that with some crude hacks (having BalancedShardsAllocator account for the different node types in earlier pathes of the code) we reduced full allocation to a minute and allocation of new shards to a second.

**Comment by ywelsch-t** [comment]
See https://github.com/ywelsch-t/elasticsearch/commit/3b9e5b2d06c722e41b7b743dda2e25ddad21d1ee

- With a profiler, I found that the two biggest pain points were `BalancedShardsAllocator.Balancer.balance()` and `BalancedShardsAllocator.Balancer.allocateUnassigned()`
- Both methods now only consider the nodes which have the right `node-type` for the relevant shard/index.
- The hack assumes that `node_type` references a single sub-cluster.

Output:

```
[2015-03-09 10:06:25,202][INFO ][org.elasticsearch.benchmark.cluster] [0] starting... 
[2015-03-09 10:06:25,237][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 26000
[2015-03-09 10:06:32,212][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 14000
[2015-03-09 10:06:35,434][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 12700
[2015-03-09 10:06:37,528][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 12100
[2015-03-09 10:06:39,269][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 11500
[2015-03-09 10:06:40,870][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 10900
[2015-03-09 10:06:42,787][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 10300
[2015-03-09 10:06:44,564][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 9700
[2015-03-09 10:06:46,231][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 9100
[2015-03-09 10:06:48,106][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 8500
[2015-03-09 10:06:49,904][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7900
[2015-03-09 10:06:51,635][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7300
[2015-03-09 10:06:53,041][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 7000
[2015-03-09 10:06:54,742][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6700
[2015-03-09 10:06:56,395][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6400
[2015-03-09 10:06:57,950][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 6100
[2015-03-09 10:06:59,304][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5800
[2015-03-09 10:07:00,799][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5500
[2015-03-09 10:07:02,430][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 5200
[2015-03-09 10:07:03,876][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4900
[2015-03-09 10:07:05,354][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4600
[2015-03-09 10:07:07,119][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4300
[2015-03-09 10:07:08,757][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 4000
[2015-03-09 10:07:10,183][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3700
[2015-03-09 10:07:11,694][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3400
[2015-03-09 10:07:13,301][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 3100
[2015-03-09 10:07:14,869][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2800
[2015-03-09 10:07:16,371][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2500
[2015-03-09 10:07:17,859][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 2200
[2015-03-09 10:07:19,440][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1900
[2015-03-09 10:07:21,087][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1600
[2015-03-09 10:07:22,607][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1300
[2015-03-09 10:07:24,130][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 1000
[2015-03-09 10:07:25,709][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 700
[2015-03-09 10:07:27,378][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 400
[2015-03-09 10:07:28,882][INFO ][org.elasticsearch.benchmark.cluster] [0] remaining unassigned 200
[2015-03-09 10:07:29,756][INFO ][org.elasticsearch.benchmark.cluster] [0] took 1m
[2015-03-09 10:07:29,756][INFO ][org.elasticsearch.benchmark.cluster] total took 1m, AVG 1m
```

**Comment by clintongormley** [comment]
@ywelsch do you know if there are still improvements to make here?

**Comment by pickypg** [comment]
@ywelsch-t ^

**Comment by ywelsch** [comment]
The remaining issue was about the BalancedShardsAllocator. I ran some tests comparing the performance of 1.5 (the version to which applied the remaining issue) to 1.7 and 2.x. On the test code I introduced above, 1.7 and 2.x both ran noticeably faster (around 2½ minutes instead of 15 minutes for 1.5). With improvements in #15678, this was reduced to 1½ minutes. The rebalance step for hot-warm setup with many nodes was also greatly improved with #15678 (taking 1 second instead of 10).

I am closing this issue as I think all points are addressed.
